trick_source/sim_services/MonteCarloGeneration/mc_variable_random_normal.cc

```diff
@@ -0,0 +1,267 @@
+     double min,
+     double max,
+     TruncationType truncType)
+{
+  truncate_low(min, truncType);
+  truncate_high(max, truncType);
+}
+
+/*****************************************************************************
+truncate_low
+Purpose:(Specifies the lower-bound of the truncation)
+Note - min is often -- but not necessarily -- a negative value.
+*****************************************************************************/
+void
+MonteCarloVariableRandomNormal::truncate_low(
+    double min,
+    TruncationType truncType)
+{
+  switch (truncType) {
+   case StandardDeviation:
+    min_value = distribution.mean() + distribution.stddev() * min;
+    break;
+   case Relative:
+    min_value = distribution.mean() + min;
+    break;
+   case Absolute:
+    min_value = min;
+    break;
+   // Unreachable code.  All types are covered.
+   default:
+     std::string message = 
+       std::string("File: ") + __FILE__ + ", Line " +
+       std::to_string(__LINE__) + ", Invalid TruncationType\nInvalid " +
+       "truncation type passed to truncate_low for variable " + 
+       variable_name.c_str() + ".\nMinimum will not be applied.\n";
+     message_publish(MSG_ERROR, message.c_str());
+    return;
+  }
+  truncated_low = true;
+}
+/*****************************************************************************
+truncate_high
+Purpose:(Specifies the upper-bound of the truncation)
+*****************************************************************************/
+void
+MonteCarloVariableRandomNormal::truncate_high( double max,
+                                               TruncationType truncType)
+{
+  switch (truncType) {
+   case StandardDeviation:
+    max_value = distribution.mean() + distribution.stddev() * max;
+    break;
+   case Relative:
+    max_value = distribution.mean() + max;
+    break;
+   case Absolute:
+    max_value = max;
+    break;
+   // Unreachable code.  All types are covered.
+   default:
+     std::string message = 
+      std::string("File: ") + __FILE__ + ", Line: " + 
+      std::to_string(__LINE__) + ", Invalid TruncationType\nInvalid " +
+      "truncation type passed to truncate_high for variable " + 
+      variable_name.c_str() + ".\nMaximum will not be applied.\n";
+      message_publish(MSG_ERROR, message.c_str());
+    return;
+  }
+  truncated_high = true;
+}
+
+/*****************************************************************************
+untruncate
+Purpose:(Remove truncation flags.)
+*****************************************************************************/
+void
+MonteCarloVariableRandomNormal::untruncate()
+{
+  truncated_low = false;
+  truncated_high = false;
+}
+
+/*****************************************************************************
+ * is_near_equal
+ * Purpose: Utility function providing double==double equality checks
+ * **************************************************************************/
+bool MonteCarloVariableRandomNormal::is_near_equal( double val1, double val2)
```

Comment:
FYI @jmpenn this is the function in question, originally written by Gary to prevent FPEs from terminating the sim during the check.  I slightly altered the `fe` system calls from the original so that we are using functions that exist  on both Mac OSX and linux.

If you have questions on the precise types of FPEs that could be thrown, Gary might have some insight into that. Yesterday in the Tricklab John had mentioned that if only underflows are of concern, most OSs automatically translate those to zero. Ultimately we just want a safe "are these two doubles equivalent?" check somewhere in the Trick framework. If y'all have a better function for us to use, we can easily change the model to point to that, just let us know.

Reply:
I think the reason that FP_underflows are being being generated is on line 256
For very tiny numbers it is generating FP_subnormals which result in FP_overflow exceptions.

From : https://www.gnu.org/software/libc/manual/html_node/FP-Exceptions.html
```
 ‘Underflow’
 The underflow exception is raised when an intermediate result
 is too small to be calculated accurately, or if the operation’s
 result rounded to the destination precision is too small to be normalized.

When no trap is installed for the underflow exception, underflow is
signaled (via the underflow flag) only when both tininess and loss
of accuracy have been detected. If no trap handler is installed the
operation continues with an imprecise small value, or zero if the
destination precision cannot hold the small exact result.
```

Some OS'es have handlers installed that set the result to 0.0. Others don't. 

For a general Trick "close_enough" function, I'd prefer to go with something that avoids FP_subnormals altogether. Below is one I wrote that avoid FP_subnormals. Note that it doesn't use subtraction to do the test : ```if ((fmin(A,B) + tolerance) >= fmax(A,B)) {``` . This is because even if the arguments are both FP_NORMAL, a subtraction can create an FP_SUBNORMAL. Example: 1.5*DBL_MIN - DBL_MIN.


```c
#include <float.h>
#include <math.h>
#include <stdio.h>

int close_enough( double A, double B, double tolerance) {

    // If an arg is a FP_SUBNORMAL that is: less than DBL_MIN (2.22507e-308)
    // then set it to zero.
    if ( fpclassify(A) == FP_SUBNORMAL ) { A = 0.0; }
    if ( fpclassify(B) == FP_SUBNORMAL ) { B = 0.0; }

    // If A and B are identical, then they're close_enough.
    if (A==B) { return true; }

    // The tolerance must be an FP_NORMAL. Nothing else makes sense.
    // Not even FP_SUBNORMAL or FP_ZERO, given the previous tests.
    // Also, a negative tolerance is silly.
    if (( fpclassify(tolerance) != FP_NORMAL ) || (tolerance < 0.0)) { return false; }

    // If the tolerance were less than the gap around the arg with
    // largest magnitude, then the args couldn't possibly be close enough.
    // So, the tolerance must be greater than or equal to the larger of
    // the gaps around A and B.
    if ( tolerance >= DBL_EPSILON * fmax( fabs(A), fabs(B))) {

        // We want to avoid using subtraction in determining
        // whether A and B are within tolerance of each other.
        // The reason is that it might generate a FP_SUBNORMAL result,
        // which would raise an FP_underflow exception. For example:
        // If A = 1.5*DBL_MIN, and B = DBL_MIN, the difference is an
        // FP_SUBNORMAL value. When FP_SUBNORMAL values are generated,
        // so are FP_underflows. By insisting that our args and tolerance
        // are not FP_SUBNORMAL, and by only using addition and comparison we
        // can avoid FP_SUBNORMAL results.
        if ((fmin(A,B) + tolerance) >= fmax(A,B)) {
            return true ; // then they're close enough.
        } else {
            return false ;
        }
    } else {
        return false;
    }
}
```